server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -576,9 +595,14 @@ private DataSegment mergeAndPush(final SegmentIdentifier identifier, final Sink
           tuningConfig.getIndexSpec()
       );
 
-      DataSegment segment = dataSegmentPusher.push(
-          mergedFile,
-          sink.getSegment().withDimensions(IndexMerger.getMergedDimensionsFromQueryableIndexes(indexes))
+      // Retry pushing segments because uploading to deep storage might fail especially for cloud storage types
+      final DataSegment segment = RetryUtils.retry(
```

Comment:
I wonder if the retrying the push should be the responsibility of the pusher, rather than the Appenderator. S3DataPusher already does its own retries, so the retries here are not really helping anything for S3. The pushers are also in a position to potentially be smarter about retries (the S3 one won't retry "access denied" for example; see `S3Utils.retryS3Operation`; so it can fail faster when retrying would be useless).

What do you think?